numpy/core/src/multiarray/refcount.c

```diff
@@ -274,15 +328,32 @@ _fillobject(char *optr, PyObject *obj, PyArray_Descr *dtype)
             _fillobject(optr + offset, obj, new);
         }
     }
-    else {
-        npy_intp i;
-        npy_intp nsize = dtype->elsize / sizeof(obj);
+    else if (PyDataType_HASSUBARRAY(dtype)) {
+        int size, i, inner_elsize;
+
+        inner_elsize = dtype->subarray->base->elsize;
+        /* Subarrays are always contiguous in memory */
+        size = dtype->elsize / inner_elsize;
 
-        for (i = 0; i < nsize; i++) {
-            Py_XINCREF(obj);
-            NPY_COPY_PYOBJECT_PTR(optr, &obj);
-            optr += sizeof(obj);
+        if (dtype->subarray->base->type_num == NPY_OBJECT) {
+            /* fast path when subarray only includes objects */
+            for (i = 0; i < size; i++){
+                Py_XINCREF(obj);
```

Comment:
Same, surely `obj` cannot be `NULL` (realize you probably copied from above, but it seems very odd).

Reply:
I think you are right in principle. But IIRC quite a few, maybe even all, internal functions actually accept interpreting `NULL` as `None` for numpy arrays. In practice I do not think numpy itself will ever fill an array with `NULL` though, so the question would be if some extensions might actually rely on being able to fill arrays with `NULL`.